wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java

```diff
@@ -310,13 +314,67 @@ public WorkspaceImpl updateWorkspace(String id, Workspace update) throws Conflic
         workspace.setConfig(new WorkspaceConfigImpl(update.getConfig()));
         update.getAttributes().put(UPDATED_ATTRIBUTE_NAME, Long.toString(currentTimeMillis()));
         workspace.setAttributes(update.getAttributes());
         return normalizeState(workspaceDao.update(workspace));
     }
 
     /**
+     * Updates workspace snapshots corresponding to update environments.
+     * Snapshots are deleted on any environment update except rename.
+     * Also they will be deleted if it is impossible to find one correspondence between old and new environment,
+     * but if in current and in update configuration the same environment is present, it supposed as untouched.
+     *
+     * @param workspaceId
+     *         id of workspace under update
+     * @param current
+     *         current workspace configuration
+     * @param update
+     *         new workspace configuration
+     */
+    private void updateWorkspaceSnapshots(String workspaceId, Workspace current, Workspace update) {
+        // this is needed to convert DTO into EnvironmentImpl object, otherwise equals will fail
+        final Map<String, Environment> updateEnvironments = update.getConfig()
+                                                                  .getEnvironments()
+                                                                  .entrySet()
+                                                                  .stream()
+                                                                  .collect(Collectors.toMap(Map.Entry::getKey,
+                                                                                            env -> new EnvironmentImpl(env.getValue())));
+
+        final Map <String, Environment> currentDiff = Sets.difference(current.getConfig().getEnvironments().entrySet(),
+                                                                      updateEnvironments.entrySet())
+                                                          .stream()
+                                                          .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+        final Map <String, Environment> updateDiff = Sets.difference(updateEnvironments.entrySet(),
+                                                                     current.getConfig().getEnvironments().entrySet())
+                                                         .stream()
+                                                         .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        for (Map.Entry<String, Environment> env : currentDiff.entrySet()) {
+            Environment envValue = env.getValue();
+            if (Collections.frequency(currentDiff.values(), envValue) == 1 &&
+                Collections.frequency(updateDiff.values(), envValue) == 1 &&
+                !updateDiff.containsKey(env.getKey())) {
+                try {
+                    updateEnvironmentSnapshots(workspaceId, env.getKey(), updateDiff.entrySet()
```

Comment:
Yes, it is possible, but I think that it will make the code less readable and ugly.
As about performance, usually we have small number of environments, so it won't be a problem. In case of only one environment this code won't be invoked.